src/solvers/refinement/string_constraint_generator_concat.cpp

```diff
@@ -57,26 +73,54 @@ string_exprt string_constraint_generatort::add_axioms_for_concat(
 
 Function: string_constraint_generatort::add_axioms_for_concat
 
+  Inputs: two string expressions
+
+ Outputs: a new string expression
+
+ Purpose: add axioms to say that the returned string expression is equal to
+          the concatenation of the two string expressions given as input
+
+\*******************************************************************/
+
+string_exprt string_constraint_generatort::add_axioms_for_concat(
+  const string_exprt &s1, const string_exprt &s2)
+{
+  exprt index_zero=from_integer(0, s2.length().type());
+  return add_axioms_for_concat_substr(s1, s2, index_zero, s2.length());
+}
+
+/*******************************************************************\
+
+Function: string_constraint_generatort::add_axioms_for_concat
+
   Inputs: function application with two arguments which are strings
 
  Outputs: a new string expression
 
  Purpose: add axioms to say that the returned string expression is equal to
           the concatenation of the two string arguments of
-          the function application
+          the function application.
+          In case 4 arguments instead of 2 are given the last two arguments
+          are intepreted as a start index and last index from which we extract
+          a substring of the second argument: this is similar to the
+          Java StringBuiledr.append(CharSequence s, int start, int end) method.
 
 \*******************************************************************/
 
 string_exprt string_constraint_generatort::add_axioms_for_concat(
   const function_application_exprt &f)
 {
   const function_application_exprt::argumentst &args=f.arguments();
-  assert(args.size()==2);
-
+  assert(args.size()>=2);
   string_exprt s1=get_string_expr(args[0]);
   string_exprt s2=get_string_expr(args[1]);
-
-  return add_axioms_for_concat(s1, s2);
+  if(args.size()!=2)
+  {
+    assert(args.size()==4);
+    return add_axioms_for_concat_substr(s1, s2, args[2], args[3]);
+  }
+  else
```

Comment:
@peterschrammel  In this case, the semantics of the code are the same with or without the `else` statement, so does it really have to go?